In [7]:
import pandas as pd
import numpy as np
import regex as re
import json

## Preparation of document format from sentence format

In [2]:
# Import json
with open("Macocu-sl-en.json", "r") as file:
	tus_content = json.load(file)

In [ ]:
# Convert data to a dataframe

corpus_df = pd.DataFrame(tus_content)

corpus_df.head()

In [4]:
# Analyze distribution
corpus_df.describe(include="all")

score_bicleaner_ai biroamer_entities translation_direction  \
count         3.176311e+06           3176311               3176311   
unique                 NaN                 2                     2   
top                    NaN                No               sl-orig   
freq                   NaN           2897662               2812317   
mean          8.809970e-01               NaN                   NaN   
std           1.299268e-01               NaN                   NaN   
min           5.000000e-01               NaN                   NaN   
25%           8.250000e-01               NaN                   NaN   
50%           9.440000e-01               NaN                   NaN   
75%           9.760000e-01               NaN                   NaN   
max           9.940000e-01               NaN                   NaN   

                                                en_source en_par_id  \
count                                             3176311   3176311   
unique                                             285892     93755   
top     http://nl.ijs.si/ME/Vault/V3/htm/mte-cesdoc-mt...     p26s0   
freq                                                 5112     36423   
mean                                                  NaN       NaN   
std                                                   NaN       NaN   
min                                                   NaN       NaN   
25%                                                   NaN       NaN   
50%                                                   NaN       NaN   
75%                                                   NaN       NaN   
max                                                   NaN       NaN   

                                                   en_par en_var_doc  \
count                                             3176311    3176311   
unique                                            2379921          4   
top     You should consider whether you understand how...          B   
freq                                                  745    1640246   
mean                                                  NaN        NaN   
std                                                   NaN        NaN   
min                                                   NaN        NaN   
25%                                                   NaN        NaN   
50%                                                   NaN        NaN   
75%                                                   NaN        NaN   
max                                                   NaN        NaN   

       en_var_dom                                          sl_source  \
count     3176311                                            3176311   
unique          4                                             456366   
top             B  http://nl.ijs.si/ME/Vault/V3/htm/mte-cesdoc-mt...   
freq      1967189                                               4294   
mean          NaN                                                NaN   
std           NaN                                                NaN   
min           NaN                                                NaN   
25%           NaN                                                NaN   
50%           NaN                                                NaN   
75%           NaN                                                NaN   
max           NaN                                                NaN   

       sl_par_id                                             sl_par  
count    3176311                                            3176311  
unique     86818                                            2542560  
top        p26s0  Izdelana je iz prijetne mešanice naravnega in ...  
freq       33594                                                207  
mean         NaN                                                NaN  
std          NaN                                                NaN  
min          NaN                                                NaN  
25%          NaN        

#### Bi-Cleaner Scores on sentence level

As we can see from the column on score_bicleaner_ai, it seems that there are no sentences with score lower than 0.5 - they must have been already filtered out? (?). We can also see, that was majority of sentences were originally written in Slovene (based on translation_direction), and that 700 English sentences and 200 Slovene sentences are repeated which could indicate some errors.

In [ ]:
# Sort by english url and then by en_par_id to order the paragraphs into texts
corpus_df = corpus_df.sort_values(by = ["en_source", "en_par_id"])

corpus_df

In [8]:
# Add information about domains
domain_re=re.compile(r'^https?://(?:www\.)?(.+?)[/$]')

en_domain_list = [domain_re.search(i).group(1) for i in corpus_df.en_source.to_list()]

en_domain_list[:3]

['15.liffe.si', '15.liffe.si', '15.liffe.si']

In [ ]:
corpus_df["en_domain"] = en_domain_list

corpus_df.head(2)

In [ ]:
# Repeat with Slovene domain
sl_domain_list = [domain_re.search(i).group(1) for i in corpus_df.sl_source.to_list()]
corpus_df["sl_domain"] = sl_domain_list

corpus_df.head(2)

In [ ]:
# Add information whether the domains are the same
corpus_df["same_domains"] = np.where(corpus_df["en_domain"] == corpus_df["sl_domain"], "yes", 'no')

# Add column for domains that are different
corpus_df["different_domains"] = corpus_df["en_domain"] + " " + corpus_df["sl_domain"]

corpus_df.head(2)

In [12]:
corpus_df["same_domains"].value_counts()

yes    2347120
no      829191
Name: same_domains, dtype: int64

In [13]:
# Analyze instances from different domains
corpus_df[corpus_df["same_domains"] != "yes"]["different_domains"] .value_counts()

eur-lex.europa.eu uradni-list.si                 21956
eur-lex.europa.eu europarl.europa.eu             16475
europarl.europa.eu eur-lex.europa.eu              8431
croatiabeachacc.com privatapartmajihrvaska.si     6907
predsednik.si up-rs.si                            6439
                                                 ...  
eesc.europa.eu okolje.si                             1
eesc.europa.eu nova-gorica.si                        1
eesc.europa.eu plastika-lvp.si                       1
eesc.europa.eu e-uprava.gov.si                       1
savacommission.org posavskiobzornik.si               1
Name: different_domains, Length: 118892, dtype: int64

In [ ]:
# Calculate average bicleaner ai score based on the en_source
corpus_df["average_score"] = corpus_df["score_bicleaner_ai"].groupby(corpus_df['en_source']).transform('mean')

corpus_df.head(2)

In [ ]:
# Add to each instance from the same en_source joint text from all sentences
corpus_df["en_doc"] = corpus_df["en_par"].groupby(corpus_df['en_source']).transform(' '.join)

corpus_df.head(2)

In [ ]:
# Repeat with the Slovene text
corpus_df["sl_doc"] = corpus_df["sl_par"].groupby(corpus_df['sl_source']).transform(' '.join)
corpus_df.head(2)

In [18]:
# Keep only one example of each text
corpus_df = corpus_df.drop_duplicates("en_doc")

corpus_df.shape

(281475, 18)

In [19]:
corpus_df.describe(include="all")

score_bicleaner_ai biroamer_entities translation_direction  \
count        281475.000000            281475                281475   
unique                 NaN                 2                     2   
top                    NaN                No               sl-orig   
freq                   NaN            245570                241480   
mean              0.841095               NaN                   NaN   
std               0.143561               NaN                   NaN   
min               0.500000               NaN                   NaN   
25%               0.740000               NaN                   NaN   
50%               0.898000               NaN                   NaN   
75%               0.963000               NaN                   NaN   
max               0.993000               NaN                   NaN   

                              en_source en_par_id  \
count                            281475    281475   
unique                           281475     15104   
top     http://15.liffe.si/?lang_chg=en     p10s0   
freq                                  1     12036   
mean                                NaN       NaN   
std                                 NaN       NaN   
min                                 NaN       NaN   
25%                                 NaN       NaN   
50%                                 NaN       NaN   
75%                                 NaN       NaN   
max                                 NaN       NaN   

                                                   en_par en_var_doc  \
count                                              281475     281475   
unique                                             268451          4   
top     You should consider whether you understand how...        UNK   
freq                                                  181     155208   
mean                                                  NaN        NaN   
std                                                   NaN        NaN   
min                                                   NaN        NaN   
25%                                                   NaN        NaN   
50%                                                   NaN        NaN   
75%                                                   NaN        NaN   
max                                                   NaN        NaN   

       en_var_dom                                     sl_source sl_par_id  \
count      281475                                        281475    281475   
unique          4                                        232628     13766   
top             B  https://www.realitica.com/en/listing/2447504     p26s0   
freq       129844                                            13      9302   
mean          NaN                                           NaN       NaN   
std           NaN                                           NaN       NaN   
min           NaN                                           NaN       NaN   
25%           NaN                                           NaN       NaN   
50%           NaN                                           NaN       NaN   
75%           NaN                                           NaN       NaN   
max           NaN                                           NaN       NaN   

                                                   sl_par        en_domain  \
count                                              281475           281475   
unique                                             250662            11959   
top     Če želite izvedeti več, lahko preberete našo p...  rumenestrani.si   
freq                                                  101            17392   
mean                                                  NaN              NaN   
std                                                   NaN              NaN   
min                                                   NaN              NaN   
25%                                                   NaN              NaN   
50%                                   

In [ ]:
corpus_df.head()

In [ ]:
# Add information about length
corpus_df["en_length"] = corpus_df.en_doc.str.split().str.len()

corpus_df.head(3)

In [22]:
corpus_df.en_length.describe()

count    281475.000000
mean        258.774364
std        1384.547486
min           1.000000
25%          29.000000
50%          79.000000
75%         207.000000
max      106893.000000
Name: en_length, dtype: float64

In [ ]:
# Add information about Slovene length
corpus_df["sl_length"] = corpus_df.sl_doc.str.split().str.len()

corpus_df.head(3)

In [48]:
# Analyze Slovene length
corpus_df.sl_length.describe()

count    141066.000000
mean        453.143635
std        2290.479800
min           2.000000
25%          78.000000
50%         148.000000
75%         307.000000
max       98530.000000
Name: sl_length, dtype: float64

In [23]:
# Discard instances that have length less than the median - less than 79
corpus_df = corpus_df[corpus_df["en_length"] > 78]

corpus_df.shape

(141066, 19)

After discarding texts that have length less than the median - less than 79 words, the Slovene-English corpus consists of 141066 aligned texts.

In [24]:
corpus_df.columns

Index(['score_bicleaner_ai', 'biroamer_entities', 'translation_direction',
       'en_source', 'en_par_id', 'en_par', 'en_var_doc', 'en_var_dom',
       'sl_source', 'sl_par_id', 'sl_par', 'en_domain', 'sl_domain',
       'same_domains', 'different_domains', 'average_score', 'en_doc',
       'sl_doc', 'en_length'],
      dtype='object')

In [25]:
# Discard irrelevant columns
corpus_df = corpus_df.drop(columns = ['score_bicleaner_ai', 'en_par_id', 'en_par', 'sl_par_id', 'sl_par'])
corpus_df.head()

biroamer_entities translation_direction  \
2584979                No               sl-orig   
1212933                No               sl-orig   
598330                Yes               sl-orig   
2796982                No               sl-orig   
642105                Yes               sl-orig   

                                   en_source en_var_doc en_var_dom  \
2584979      http://15.liffe.si/?lang_chg=en          B          B   
1212933      http://16.liffe.si/?lang_chg=en          B          B   
598330       http://17.liffe.si/?lang_chg=en          B          B   
2796982  http://1proti1.mg-lj.si/en/concept/        UNK        UNK   
642105       http://2006.fdf.si/?lang_chg=en          B          B   

                                                 sl_source         en_domain  \
2584979                    http://15.liffe.si/?lang_chg=sl       15.liffe.si   
1212933       http://16.liffe.si/index.php?menu_item=domov       16.liffe.si   
598330                     http://17.liffe.si/?lang_chg=sl       17.liffe.si   
2796982  http://www.mg-lj.si/si/razstave/98/vmesna-postaja  1proti1.mg-lj.si   
642105   http://2006.fdf.si/?menu_item=podatki&amp;menu...       2006.fdf.si   

           sl_domain same_domains          different_domains  average_score  \
2584979  15.liffe.si          yes    15.liffe.si 15.liffe.si       0.936808   
1212933  16.liffe.si          yes    16.liffe.si 16.liffe.si       0.900000   
598330   17.liffe.si          yes    17.liffe.si 17.liffe.si       0.957875   
2796982     mg-lj.si           no  1proti1.mg-lj.si mg-lj.si       0.945633   
642105   2006.fdf.si          yes    2006.fdf.si 2006.fdf.si       0.941667   

                                                    en_doc  \
2584979  It went out with a bang. The evening sparkled ...   
1212933  Some days ago the organisers of the 17th Liffe...   
598330   17th LIFFe was brought to an end with the best...   
2796982  Concept Not interested in showcasing art, 1:1 ...   
642105   The 8th International Documentary Film Festiva...   

                                                    sl_doc  en_length  
2584979  Končalo se je razburljivo in z razkošjem. Veče...        601  
1212933  Pred dnevi smo se iz 59. mednarodnega filmskeg...        293  
598330   S podelitvijo nagrad in predvajanjem Režiserja...        445  
2796982  VMESNO POSTAJO 1:1 organiziramo v koprodukciji...        827  
642105   S slovenskim filmom Poročno potovanje se je ko...        423

In [26]:
# Inspect corpus information
corpus_df.describe(include="all")

biroamer_entities translation_direction  \
count             141066                141066   
unique                 2                     2   
top                   No               sl-orig   
freq              122869                124561   
mean                 NaN                   NaN   
std                  NaN                   NaN   
min                  NaN                   NaN   
25%                  NaN                   NaN   
50%                  NaN                   NaN   
75%                  NaN                   NaN   
max                  NaN                   NaN   

                              en_source en_var_doc en_var_dom  \
count                            141066     141066     141066   
unique                           141066          4          4   
top     http://15.liffe.si/?lang_chg=en        UNK          B   
freq                                  1      55840      77117   
mean                                NaN        NaN        NaN   
std                                 NaN        NaN        NaN   
min                                 NaN        NaN        NaN   
25%                                 NaN        NaN        NaN   
50%                                 NaN        NaN        NaN   
75%                                 NaN        NaN        NaN   
max                                 NaN        NaN        NaN   

                                                sl_source    en_domain  \
count                                              141066       141066   
unique                                             126398         8507   
top     https://www.sofascore.com/sl/ekipa/nogomet/vik...  oblacila.si   
freq                                                    9         3982   
mean                                                  NaN          NaN   
std                                                   NaN          NaN   
min                                                   NaN          NaN   
25%                                                   NaN          NaN   
50%                                                   NaN          NaN   
75%                                                   NaN          NaN   
max                                                   NaN          NaN   

          sl_domain same_domains        different_domains  average_score  \
count        141066       141066                   141066  141066.000000   
unique        11529            2                    25040            NaN   
top     oblacila.si          yes  oblacila.si oblacila.si            NaN   
freq           3952       104853                     3949            NaN   
mean            NaN          NaN                      NaN       0.875767   
std             NaN          NaN                      NaN       0.074288   
min             NaN          NaN                      NaN       0.502000   
25%             NaN          NaN                      NaN       0.833667   
50%             NaN          NaN                      NaN       0.894077   
75%             NaN          NaN                      NaN       0.932000   
max             NaN          NaN                      NaN       0.990500   

                                                   en_doc  \
count                                              141066   
unique                                             141066   
top     It went out with a bang. The evening sparkled ...   
freq                                                    1   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                                   sl_doc      en_length  
count          

In [55]:
# Inspect en_var_doc statistics
print(corpus_df.en_var_doc.value_counts(normalize = True).to_markdown())

|     |   en_var_doc |
|:----|-------------:|
| B   |    0.390137  |
| UNK |    0.389793  |
| A   |    0.162199  |
| MIX |    0.0578715 |


In [57]:
print(corpus_df.en_var_dom.value_counts(normalize = True).to_markdown())

|     |   en_var_dom |
|:----|-------------:|
| B   |   0.539765   |
| MIX |   0.284293   |
| A   |   0.16629    |
| UNK |   0.00965161 |


In [58]:
# Inspect translation direction
print(corpus_df.translation_direction.value_counts(normalize = True).to_markdown())

|         |   translation_direction |
|:--------|------------------------:|
| sl-orig |                0.886241 |
| en-orig |                0.113759 |


In [60]:
print(corpus_df.average_score.describe().to_markdown())

|       |   average_score |
|:------|----------------:|
| count |  104853         |
| mean  |       0.886736  |
| std   |       0.0682533 |
| min   |       0.502     |
| 25%   |       0.8519    |
| 50%   |       0.904     |
| 75%   |       0.93675   |
| max   |       0.9905    |


#### Main information on the corpus

As we can see, almost all of the documents were originally written in Slovene (88%). Most of them are identified as "UNK" (40%), followed by British (39%), and much less American texts, by English variety detection on document level. On the domain level, most of them (55%) were identified to be British. Most of the texts have quality higher than 0.90 based on the bicleaner score. There are 23 Slovene duplicated texts.

In [40]:
# Analyze instances from different domains
corpus_df["same_domains"].value_counts()

yes    104853
no      36213
Name: same_domains, dtype: int64

In [41]:
corpus_df[corpus_df["same_domains"] != "yes"]["different_domains"] .value_counts()

eugo.gov.si spot.gov.si                                           366
plus500.com plus500.si                                            283
croatiabeachacc.com privatapartmajihrvaska.si                     281
hajdi.net hajdi.si                                                258
bsi.si bsi.azurewebsites.net                                      257
                                                                 ... 
etnamaar.com casmatino.si                                           1
etnologija.etnoinfolab.org anglistika.ff.uni-lj.si                  1
etnologija.etnoinfolab.org ffa.uni-lj.si                            1
etnologija.etnoinfolab.org pedagogika-andragogika.ff.uni-lj.si      1
塞尔维亚.realigro.asia oglasi.si                                        1
Name: different_domains, Length: 19060, dtype: int64

In [51]:
len(corpus_df.different_domains.unique())

25040

In [ ]:
# Inspect duplicated Slovene texts
pd.set_option('display.max_colwidth', None)
duplicated = corpus_df[corpus_df.sl_doc.duplicated(keep=False)].sort_values("sl_doc")
duplicated[["sl_doc", "en_doc", "average_score"]]

In [50]:
# Inspect a random sample
pd.set_option('display.max_colwidth', 300)
corpus_df.sample(n=20)

biroamer_entities translation_direction  \
1548798               Yes               sl-orig   
84755                 Yes               sl-orig   
2562108                No               sl-orig   
348403                 No               sl-orig   
2031466                No               sl-orig   
609292                 No               sl-orig   
1805773                No               sl-orig   
1737177                No               sl-orig   
1993535                No               sl-orig   
930061                 No               sl-orig   
2223005                No               sl-orig   
71124                  No               sl-orig   
245429                Yes               en-orig   
806658                 No               sl-orig   
1126067                No               sl-orig   
2598465                No               sl-orig   
577848                 No               sl-orig   
739258                 No               sl-orig   
1545007                No               sl-orig   
1668855                No               sl-orig   

                                                                                                                        en_source  \
1548798                                        https://www.bled.si/en/what-to-see-do/cuisine/2019100208524126/educational-beehive   
84755                                                                                         http://pefprints.pef.uni-lj.si/469/   
2562108                                                               http://newsletter.etwinning.net/en/pub/2013/etnl_062013.htm   
348403                                  https://oblacila.si/under-armour-under-armour-tricko-performance-polo-2.0-blu__7wyPxjjEtG   
2031466                                                             http://eugo.gov.si/en/activities/activity/14091/showActivity/   
609292                                                                    https://www.pokerharder.com/en-gb/learn-poker/variants/   
1805773                                                      http://visitcerkno.si/en/boulders-on-the-crni-vrh-hill-above-cerkno/   
1737177                                                                                    https://www.nuk.uni-lj.si/eng/node/489   
1993535                                      https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=LEGISSUM:l28056&amp;from=HR   
930061                       https://www.agencia.si/en/print/243524-oddaja-poslovni-prostor-skladisce-obalno-kraska-sezana-storje   
2223005             http://arhivffweb2020.ff.uni-lj.si/an/study/departments/department_comparative_literature_and_literary_theory   
71124                           https://www.mirovni-institut.si/en/projects/informal-reproductive-work-trends-in-slovenia-and-eu/   
245429                                                 http://ricinus2.mf.uni-lj.si/bs/neva.exe?name=bs_cx&amp;e=%22Vancomycin%22   
806658                                                                        https://www.spiderfestival.com/tandem-kresnicka-eng   
1126067                                                                           https://arhiv.akos-rs.si/railway-service-market   
2598465                                                                         https://www.bekament.com/en/proizvodi/bk-podloga/   
577848                                                              https://www.consilium.europa.eu/en/policies/eu-recovery-plan/   
739258                                                   http://www.cer-slo.si/1green-breakfast---greentech-forum-14-02-2019.html   
1545007                                                                                         https://roglab.si/en/news/item-0/   
1668855  http://www.sayhellototheworld.eu/en/gallery/year-2016-2017/first-call/first-video-call-warsaw-slovenia-vilnius-lithuania   

        en_var_doc en_var_dom  \
1548798          B          B   
84755            A          B   
2562108          B          B   
348403    

In [52]:
# Discard instances document that are from different domains
corpus_df = corpus_df[corpus_df["same_domains"] == "yes"]
corpus_df.describe(include="all")

biroamer_entities translation_direction  \
count             104853                104853   
unique                 2                     2   
top                   No               sl-orig   
freq               91558                 92925   
mean                 NaN                   NaN   
std                  NaN                   NaN   
min                  NaN                   NaN   
25%                  NaN                   NaN   
50%                  NaN                   NaN   
75%                  NaN                   NaN   
max                  NaN                   NaN   

                              en_source en_var_doc en_var_dom  \
count                            104853     104853     104853   
unique                           104853          4          4   
top     http://15.liffe.si/?lang_chg=en          B          B   
freq                                  1      40907      56596   
mean                                NaN        NaN        NaN   
std                                 NaN        NaN        NaN   
min                                 NaN        NaN        NaN   
25%                                 NaN        NaN        NaN   
50%                                 NaN        NaN        NaN   
75%                                 NaN        NaN        NaN   
max                                 NaN        NaN        NaN   

                                                             sl_source  \
count                                                           104853   
unique                                                           95040   
top     https://www.sofascore.com/sl/ekipa/nogomet/viktoria-plzen/4502   
freq                                                                 9   
mean                                                               NaN   
std                                                                NaN   
min                                                                NaN   
25%                                                                NaN   
50%                                                                NaN   
75%                                                                NaN   
max                                                                NaN   

          en_domain    sl_domain same_domains        different_domains  \
count        104853       104853       104853                   104853   
unique         5980         5980            1                     5980   
top     oblacila.si  oblacila.si          yes  oblacila.si oblacila.si   
freq           3949         3949       104853                     3949   
mean            NaN          NaN          NaN                      NaN   
std             NaN          NaN          NaN                      NaN   
min             NaN          NaN          NaN                      NaN   
25%             NaN          NaN          NaN                      NaN   
50%             NaN          NaN          NaN                      NaN   
75%             NaN          NaN          NaN                      NaN   
max             NaN          NaN          NaN                      NaN   

        average_score  \
count   104853.000000   
unique            NaN   
top               NaN   
freq              NaN   
mean         0.886736   
std          0.068253   
min          0.502000   
25%          0.851900   
50%          0.904000   
75%          0.936750   
max          0.990500   

                                                                                                                                                                                                                                                                                                             en_doc  \
count                                                                                                                                                                                                                                      

In [53]:
# Save the dataframe to csv
corpus_df.to_csv("Macocu-sl-en-doc-format.csv", sep= "\t")